extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoLinkDiscoveryService.java

```diff
@@ -0,0 +1,274 @@
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.CharacterData;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+
+/**
+ * The {@link WemoLinkDiscoveryService} is responsible for discovering new and
+ * removed WeMo devices connected to the WeMo Link Bridge.
+ *
+ * @author Hans-Jörg Merk - Initial contribution
+ *
+ */
+public class WemoLinkDiscoveryService extends AbstractDiscoveryService implements UpnpIOParticipant {
+
+    private Logger logger = LoggerFactory.getLogger(WemoLinkDiscoveryService.class);
+
+    public final static Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_MZ100);
+
+    /**
+     * Maximum time to search for devices in seconds.
+     */
+    private final static int SEARCH_TIME = 20;
+
+    /**
+     * Initial delay for scanning job in seconds.
+     */
+    private final static int INITIAL_DELAY = 5;
+
+    /**
+     * Scan interval for scanning job in seconds.
+     */
+    private final static int SCAN_INTERVAL = 120;
+
+    private WemoBridgeHandler wemoBridgeHandler;
+
+    /**
+     * Job which will do the background scanning
+     */
+    private WemoLinkScan scanningRunnable;
+    /**
+     * Schedule for scanning
+     */
+    private ScheduledFuture<?> scanningJob;
+
+    private UpnpIOService service;
+
+    public WemoLinkDiscoveryService(WemoBridgeHandler wemoBridgeHandler, UpnpIOService upnpIOService) {
+        super(SEARCH_TIME);
+        this.wemoBridgeHandler = wemoBridgeHandler;
+
+        if (upnpIOService != null) {
+            this.service = upnpIOService;
+        } else {
+            logger.debug("upnpIOService not set.");
+        }
+
+        this.scanningRunnable = new WemoLinkScan(this);
+        if (wemoBridgeHandler == null) {
+            logger.warn("no bridge handler for scan given");
+        }
+        this.activate(null);
+    }
+
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+        return SUPPORTED_THING_TYPES;
+    }
+
+    @Override
+    protected void startScan() {
+
+        logger.trace("Starting WeMoEndDevice discovery on WeMo Link {}", wemoBridgeHandler.getThing().getUID());
+        try {
+
+            String devUDN = wemoBridgeHandler.getThing().getConfiguration().get(ROOT_UDN).toString();
+            String soapHeader = "\"urn:Belkin:service:bridge:1#GetEndDevices\"";
+            String content = "<?xml version=\"1.0\"?>"
+                    + "<s:Envelope xmlns:s=\"http://schemas.xmlsoap.org/soap/envelope/\" s:encodingStyle=\"http://schemas.xmlsoap.org/soap/encoding/\">"
+                    + "<s:Body>" + "<u:GetEndDevices xmlns:u=\"urn:Belkin:service:bridge:1\">" + "<DevUDN>" + devUDN
+                    + "</DevUDN><ReqListType>PAIRED_LIST</ReqListType>" + "</u:GetEndDevices>" + "</s:Body>"
+                    + "</s:Envelope>";
+
+            URL descriptorURL = service.getDescriptorURL(this);
```

Comment:
You are accepting NULL for the upnpIOService, just logging it on debug level. In that case, I see a potential NPE here